### PR TITLE
Leaf 1830 - send back comment

### PR DESCRIPTION
**Pull request by jampaul3**
This allows an admin to select to require a comment when sending back to requestor. If checked then it will require a comment, if no comment added a popup will inform that a comment is needed.

Side project, I added 2 new methods to the Db class. pdo_select_query and pdo_update_query. These now return a properly formatted json response. The select will also send back the data being requested, the update does not because an update query does not have any data to send back. This also sends back a json response if the query failed for some reason, as well as if there is a PDO Exception. The Exception also writes to the error_log, so hopefully that will help us in the future track down sql errors.

### CODE REVIEW

File: LEAF_Request_Portal/js/workflow.js [hunk 1]
```diff
@@ -5,734 +5,805 @@ var workflow;
...
-          }
-        };
+        }
 
-        if (actionPreconditionFunc !== undefined) {
-          actionPreconditionFunc(e.data, completeAction);
-        } else {
-          completeAction();
+        // legacy workflow modules based on dependencyIDs
+        for (var u in step.jsSrcList) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/js/workflow.js:420]: var to let

File: LEAF_Request_Portal/js/workflow.js [hunk 2]
```diff
@@ -5,734 +5,805 @@ var workflow;
...
+                        $("#" + containerID).append(
+                            '<div id="workflowbox_lastAction" class="workflowbox" style="padding: 0px; margin-top: 8px"></div>'
+                        );
+                        $("#workflowbox_lastAction").css({
+                            "background-color": response.stepBgColor,
+                            border: response.stepBorder,
+                        });
+                    }
+
+                    var date = new Date(response.time * 1000);
```
Comment (Pelentan) [inline LEAF_Request_Portal/js/workflow.js:586]: Do all these need to be var's?  A lot of them look like ones only used locally.

File: LEAF_Request_Portal/js/workflow.js [hunk 3]
```diff
@@ -5,734 +5,805 @@ var workflow;
...
+    /**
+     * @memberOf LeafWorkflow
+     */
+    function getWorkflow(recordID) {
+        $("#" + containerID).empty();
+        $("#" + containerID).css("display", "none");
+        antiDblClick = 0;
+        currRecordID = recordID;
+
+        var masquerade = "";
```
Comment (Pelentan) [inline LEAF_Request_Portal/js/workflow.js:749]: var to let?

File: LEAF_Request_Portal/js/workflow.js [hunk 4]
```diff
@@ -5,734 +5,805 @@ var workflow;
...
-    var masquerade = "";
-    if (window.location.href.indexOf("masquerade=nonAdmin") != -1) {
-      masquerade = "?masquerade=nonAdmin";
+                "api/formWorkflow/" +
+                recordID +
+                "/currentStep" +
+                masquerade,
+            dataType: "json",
+            success: function (res) {
+                for (var i in res) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/js/workflow.js:764]: var to let

**Review by Pelentan — CHANGES_REQUESTED** [review]

File: LEAF_Request_Portal/js/workflow.js (hunk 1 above)
Comment (jampaul3) [reply LEAF_Request_Portal/js/workflow.js:420]: I'll make these changes but I think they only showed up because of some formatting changes.

**Review by raphaelfretter — APPROVED** [review]

**Review by jrodcode — APPROVED** [review]

**Review by Pelentan — APPROVED** [review]

**Review by shaneodd — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by AmandaSumner — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Pull request merged**